filters/FaceRasterFilter.cpp

```diff
@@ -0,0 +1,142 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+ * OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+ * OF SUCH DAMAGE.
+ ****************************************************************************/
+
+#include "FaceRasterFilter.hpp"
+
+#include <pdal/util/Utils.hpp>
+#include <pdal/private/MathUtils.hpp>
+
+namespace pdal
+{
+
+static StaticPluginInfo const s_info
+{
+    "filters.faceraster",
+    "Face Raster Filter",
+    "http://pdal.io/stages/filters.faceraster.html"
+};
+
+CREATE_STATIC_STAGE(FaceRasterFilter, s_info)
+
+std::string FaceRasterFilter::getName() const
+{
+    return s_info.name;
+}
+
+
+FaceRasterFilter::FaceRasterFilter()
+{}
+
+
+void FaceRasterFilter::addArgs(ProgramArgs& args)
+{
+    m_limits.addArgs(args);
+    args.add("mesh", "Mesh name", m_meshName);
+}
+
+void FaceRasterFilter::prepared(PointTableRef)
+{
+    int cnt = m_limits.checkArgs();
+    if (cnt != 0 && cnt != 4)
+        throwError("Must specify all or none of 'origin_x', 'origin_y', 'width' and 'height'.");
+    m_computeLimits = (cnt == 0);
+}
+
+
+void FaceRasterFilter::filter(PointView& v)
+{
+    // If the user hasn't set bounds, set them based on the data.
+    if (!m_computeLimits)
+    {
+        BOX2D bounds;
+        v.calculateBounds(bounds);
+        m_limits.xOrigin = bounds.minx - (m_limits.edgeLength / 2);
+        m_limits.yOrigin = bounds.miny - (m_limits.edgeLength / 2);
+        m_limits.width = ((bounds.maxx - m_limits.xOrigin) / m_limits.edgeLength) + 1;
+        m_limits.height = ((bounds.maxy - m_limits.yOrigin) / m_limits.edgeLength) + 1;
+    }
+    Rasterd *raster = v.createRaster("faceraster", m_limits);
+    if (!raster)
+        throwError("Raster already exists");
+
+    TriangularMesh *m = v.mesh(m_meshName);
+    if (!m)
+        throwError("Mesh '" + m_meshName + "; does not exist.");
+
+    for (const Triangle& t : *m)
+    {
+        double x1 = v.getFieldAs<double>(Dimension::Id::X, t.m_a) - m_limits.xOrigin;
+        double y1 = v.getFieldAs<double>(Dimension::Id::Y, t.m_a) - m_limits.yOrigin;
+        double z1 = v.getFieldAs<double>(Dimension::Id::Z, t.m_a);
+
+        double x2 = v.getFieldAs<double>(Dimension::Id::X, t.m_b) - m_limits.xOrigin;
+        double y2 = v.getFieldAs<double>(Dimension::Id::Y, t.m_b) - m_limits.yOrigin;
+        double z2 = v.getFieldAs<double>(Dimension::Id::Z, t.m_b);
+
+        double x3 = v.getFieldAs<double>(Dimension::Id::X, t.m_c) - m_limits.xOrigin;
+        double y3 = v.getFieldAs<double>(Dimension::Id::Y, t.m_c) - m_limits.yOrigin;
+        double z3 = v.getFieldAs<double>(Dimension::Id::Z, t.m_c);
+
+        double xmax = (std::max)((std::max)(x1, x2), x3);
+        double xmin = (std::min)((std::min)(x1, x2), x3);
+        double ymax = (std::max)((std::max)(y1, y2), y3);
+        double ymin = (std::min)((std::min)(y1, y2), y3);
+
+        double halfEdge = m_limits.edgeLength / 2;
```

Comment:
This _could_ be moved to the top of `filter()` and `halfEdge` used when setting `xOrigin` and `yOrigin`. Also, no need to recompute for each triangle. Doubt it makes any noticeable difference to do so though.

Reply:
I expect the compiler would do this for me, but I just missed this. Thx.